File: benchmarks/v3/scenarios/05-java-saga/with-mcp/test/OrderFulfillmentOrchestrator.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import {
  Order,
  OrderItem,
  SagaContext,
  SagaError,
  OrderService,
  InventoryService,
  PaymentService,
  ShippingService
} from '../domain';
import { OrderFulfillmentOrchestrator } from '../application/saga/OrderFulfillmentOrchestrator';
import { CreateOrderStep } from '../application/saga/CreateOrderStep';
import { ReserveInventoryStep } from '../application/saga/ReserveInventoryStep';
import { ProcessPaymentStep } from '../application/saga/ProcessPaymentStep';
import { ShipOrderStep } from '../application/saga/ShipOrderStep';

describe('OrderFulfillmentOrchestrator', () => {
  let orchestrator: OrderFulfillmentOrchestrator;
  let orderService: OrderService;
  let inventoryService: InventoryService;
  let paymentService: PaymentService;
  let shippingService: ShippingService;

  const testItems: OrderItem[] = [
    { productId: 'prod-1', quantity: 2, unitPrice: 50 }
  ];

  const testOrder = Order.create({
    id: 'order-123',
    customerId: 'customer-456',
    items: testItems,
    totalAmount: 100
  });

  beforeEach(() => {
    // Create mock services
    orderService = {
      createOrder: vi.fn().mockResolvedValue(testOrder),
      cancelOrder: vi.fn().mockResolvedValue(undefined),
      getOrder: vi.fn().mockResolvedValue(testOrder)
    };

    inventoryService = {
      reserveInventory: vi.fn().mockResolvedValue('reservation-789'),
      releaseInventory: vi.fn().mockResolvedValue(undefined),
      checkAvailability: vi.fn().mockResolvedValue(true)
    };

    paymentService = {
      processPayment: vi.fn().mockResolvedValue('txn-101'),
      refundPayment: vi.fn().mockResolvedValue(undefined)
    };

    shippingService = {
      createShipment: vi.fn().mockResolvedValue('tracking-202'),
      cancelShipment: vi.fn().mockResolvedValue(undefined)
    };

    // Create saga steps
    const createOrderStep = new CreateOrderStep(orderService);
    const reserveInventoryStep = new ReserveInventoryStep(inventoryService);
    const processPaymentStep = new ProcessPaymentStep(paymentService);
    const shipOrderStep = new ShipOrderStep(shippingService);

    orchestrator = new OrderFulfillmentOrchestrator([
      createOrderStep,
      reserveInventoryStep,
      processPaymentStep,
      shipOrderStep
    ]);
  });

  describe('Happy Path', () => {
    it('should_complete_saga_successfully_when_all_steps_pass', async () => {
      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(result.success).toBe(true);
      expect(result.completedSteps).toEqual([
        'CreateOrder',
        'ReserveInventory',
        'ProcessPayment',
        'ShipOrder'
      ]);
      expect(result.failedStep).toBeUndefined();
      expect(result.error).toBeUndefined();
      expect(result.context.order).toBeDefined();
      expect(result.context.inventoryReservationId).toBe('reservation-789');
      expect(result.context.paymentTransactionId).toBe('txn-101');
      expect(result.context.shipmentTrackingId).toBe('tracking-202');
    });

    it('should_track_execution_state_through_saga', async () => {
      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(orderService.createOrder).toHaveBeenCalledTimes(1);
      expect(inventoryService.reserveInventory).toHaveBeenCalledTimes(1);
      expect(paymentService.processPayment).toHaveBeenCalledTimes(1);
      expect(shippingService.createShipment).toHaveBeenCalledTimes(1);
    });
  });

  describe('Failure Scenarios with Compensation', () => {
    it('should_compensate_when_create_order_fails', async () => {
      vi.mocked(orderService.createOrder).mockRejectedValue(
        new Error('Order creation failed')
      );

      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(result.success).toBe(false);
      expect(result.failedStep).toBe('CreateOrder');
      expect(result.completedSteps).toEqual([]);
      // No compensation needed since first step failed
      expect(result.compensationResults?.completedCompensations).toEqual([]);
    });

    it('should_compensate_create_order_when_reserve_inventory_fails', async () => {
      vi.mocked(inventoryService.reserveInventory).mockRejectedValue(
        new Error('Inventory not available')
      );

      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(result.success).toBe(false);
      expect(result.failedStep).toBe('ReserveInventory');
      expect(result.completedSteps).toEqual(['CreateOrder']);
      expect(result.compensationResults?.triggered).toBe(true);
      expect(result.compensationResults?.completedCompensations).toEqual([
        'CreateOrder'
      ]);
      expect(orderService.cancelOrder).toHaveBeenCalledWith('order-123');
    });

    it('should_compensate_inventory_and_order_when_payment_fails', async () => {
      vi.mocked(paymentService.processPayment).mockRejectedValue(
        new Error('Payment declined')
      );

      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(result.success).toBe(false);
      expect(result.failedStep).toBe('ProcessPayment');
      expect(result.completedSteps).toEqual(['CreateOrder', 'ReserveInventory']);
      expect(result.compensationResults?.triggered).toBe(true);
      expect(result.compensationResults?.completedCompensations).toContain(
        'ReserveInventory'
      );
      expect(result.compensationResults?.completedCompensations).toContain(
        'CreateOrder'
      );
      expect(inventoryService.releaseInventory).toHaveBeenCalledWith(
        'reservation-789'
      );
      expect(orderService.cancelOrder).toHaveBeenCalledWith('order-123');
    });

    it('should_compensate_all_steps_when_shipping_fails', async () => {
      vi.mocked(shippingService.createShipment).mockRejectedValue(
        new Error('Shipping unavailable')
      );

      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(result.success).toBe(false);
      expect(result.failedStep).toBe('ShipOrder');
      expect(result.completedSteps).toEqual([
        'CreateOrder',
        'ReserveInventory',
        'ProcessPayment'
      ]);
      expect(result.compensationResults?.triggered).toBe(true);
      expect(result.compensationResults?.completedCompensations).toHaveLength(3);
      expect(paymentService.refundPayment).toHaveBeenCalledWith('txn-101');
      expect(inventoryService.releaseInventory).toHaveBeenCalledWith(
        'reservation-789'
      );
      expect(orderService.cancelOrder).toHaveBeenCalledWith('order-123');
    });

    it('should_handle_compensation_failure_gracefully', async () => {
      vi.mocked(shippingService.createShipment).mockRejectedValue(
        new Error('Shipping unavailable')
      );
      vi.mocked(paymentService.refundPayment).mockRejectedValue(
        new Error('Refund failed')
      );

      const context = SagaContext.create('order-123')
        .withMetadata('customerId', 'customer-456')
        .withMetadata('items', testItems)
        .withMetadata('totalAmount', 100);

      const result = await orchestrator.execute(context);

      expect(result.success).toBe(false);
      expect(result.compensationResults?.triggered).toBe(true);
      expect(result.compensationResults?.failedCompensations).toContain(
        'ProcessPayment'
      );
      // Other compensations should still complete
      expect(result.compensationResults?.completedCompensations).toContain(
        'ReserveInventory'
      );
      expect(result.compensationResults?.completedCompensations).toContain(
        'CreateOrder'
      );
    });
  });

  describe('Orchestrator Configuration', () => {
    it('should_return_registered_steps', () => {
      const steps = orchestrator.getSteps();

      expect(steps).toHaveLength(4);
      expect(steps.map((s) => s.name)).toEqual([
        'CreateOrder',
        'ReserveInventory',
        'ProcessPayment',
        'ShipOrder'
      ]);
    });
  });
});

File: benchmarks/v3/scenarios/05-java-saga/without-mcp/src/test/java/com/example/saga/saga/OrderFulfillmentOrchestratorTest.java
package com.example.saga.saga;

import com.example.saga.application.port.InventoryService;
import com.example.saga.application.port.OrderService;
import com.example.saga.application.port.PaymentService;
import com.example.saga.application.port.ShippingService;
import com.example.saga.application.saga.CompensationSummary;
import com.example.saga.application.saga.SagaContext;
import com.example.saga.application.saga.SagaExecutionResult;
import com.example.saga.application.saga.SagaStep;
import com.example.saga.application.saga.orchestrator.OrderFulfillmentOrchestrator;
import com.example.saga.application.saga.step.CreateOrderStep;
import com.example.saga.application.saga.step.ProcessPaymentStep;
import com.example.saga.application.saga.step.ReserveInventoryStep;
import com.example.saga.application.saga.step.ShipOrderStep;
import com.example.saga.domain.entity.Order;
import com.example.saga.domain.valueobject.OrderItem;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyList;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
@DisplayName("OrderFulfillmentOrchestrator")
class OrderFulfillmentOrchestratorTest {

    @Mock
    private OrderService orderService;

    @Mock
    private InventoryService inventoryService;

    @Mock
    private PaymentService paymentService;

    @Mock
    private ShippingService shippingService;

    private OrderFulfillmentOrchestrator orchestrator;

    private List<OrderItem> testItems;
    private Order testOrder;

    @BeforeEach
    void setUp() {
        testItems = Arrays.asList(
                new OrderItem("prod-1", 2, new BigDecimal("50.00"))
        );

        testOrder = Order.create(
                "order-123",
                "customer-456",
                testItems,
                new BigDecimal("100.00")
        );

        // Create saga steps
        CreateOrderStep createOrderStep = new CreateOrderStep(orderService);
        ReserveInventoryStep reserveInventoryStep = new ReserveInventoryStep(inventoryService);
        ProcessPaymentStep processPaymentStep = new ProcessPaymentStep(paymentService);
        ShipOrderStep shipOrderStep = new ShipOrderStep(shippingService);

        List<SagaStep<?>> steps = Arrays.asList(
                createOrderStep,
                reserveInventoryStep,
                processPaymentStep,
                shipOrderStep
        );

        orchestrator = new OrderFulfillmentOrchestrator(steps);
    }

    @Nested
    @DisplayName("Happy Path")
    class HappyPath {

        @BeforeEach
        void setUpMocks() {
            when(orderService.createOrder(anyString(), anyList(), any(BigDecimal.class)))
                    .thenReturn(testOrder);
            when(inventoryService.reserveInventory(anyString(), anyList()))
                    .thenReturn("reservation-789");
            when(paymentService.processPayment(anyString(), anyString(), any(BigDecimal.class)))
                    .thenReturn("txn-101");
            when(shippingService.createShipment(anyString(), anyString()))
                    .thenReturn("tracking-202");
        }

        @Test
        @DisplayName("should complete saga successfully when all steps pass")
        void shouldCompleteSagaSuccessfullyWhenAllStepsPass() {
            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            SagaExecutionResult result = orchestrator.execute(context);

            assertThat(result.isSuccess()).isTrue();
            assertThat(result.getCompletedSteps()).containsExactly(
                    "CreateOrder",
                    "ReserveInventory",
                    "ProcessPayment",
                    "ShipOrder"
            );
            assertThat(result.getFailedStep()).isEmpty();
            assertThat(result.getError()).isEmpty();
            assertThat(result.getContext().getOrder()).isPresent();
            assertThat(result.getContext().getInventoryReservationId())
                    .contains("reservation-789");
            assertThat(result.getContext().getPaymentTransactionId())
                    .contains("txn-101");
            assertThat(result.getContext().getShipmentTrackingId())
                    .contains("tracking-202");
        }

        @Test
        @DisplayName("should track execution state through saga")
        void shouldTrackExecutionStateThroughSaga() {
            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            orchestrator.execute(context);

            verify(orderService).createOrder(
                    eq("customer-456"),
                    eq(testItems),
                    eq(new BigDecimal("100.00"))
            );
            verify(inventoryService).reserveInventory(eq("order-123"), anyList());
            verify(paymentService).processPayment(
                    eq("order-123"),
                    eq("customer-456"),
                    eq(new BigDecimal("100.00"))
            );
            verify(shippingService).createShipment(eq("order-123"), eq("customer-456"));
        }
    }

    @Nested
    @DisplayName("Failure Scenarios with Compensation")
    class FailureScenarios {

        @Test
        @DisplayName("should compensate when create order fails")
        void shouldCompensateWhenCreateOrderFails() {
            when(orderService.createOrder(anyString(), anyList(), any(BigDecimal.class)))
                    .thenThrow(new RuntimeException("Order creation failed"));

            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            SagaExecutionResult result = orchestrator.execute(context);

            assertThat(result.isSuccess()).isFalse();
            assertThat(result.getFailedStep()).contains("CreateOrder");
            assertThat(result.getCompletedSteps()).isEmpty();
            // No compensation needed since first step failed
            assertThat(result.getCompensationResults())
                    .isPresent()
                    .get()
                    .extracting(CompensationSummary::getCompletedCompensations)
                    .asList()
                    .isEmpty();
        }

        @Test
        @DisplayName("should compensate create order when reserve inventory fails")
        void shouldCompensateCreateOrderWhenReserveInventoryFails() {
            when(orderService.createOrder(anyString(), anyList(), any(BigDecimal.class)))
                    .thenReturn(testOrder);
            when(inventoryService.reserveInventory(anyString(), anyList()))
                    .thenThrow(new RuntimeException("Inventory not available"));

            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            SagaExecutionResult result = orchestrator.execute(context);

            assertThat(result.isSuccess()).isFalse();
            assertThat(result.getFailedStep()).contains("ReserveInventory");
            assertThat(result.getCompletedSteps()).containsExactly("CreateOrder");
            assertThat(result.getCompensationResults())
                    .isPresent()
                    .get()
                    .extracting(CompensationSummary::isTriggered)
                    .isEqualTo(true);
            assertThat(result.getCompensationResults().get().getCompletedCompensations())
                    .containsExactly("CreateOrder");

            verify(orderService).cancelOrder(eq("order-123"));
        }

        @Test
        @DisplayName("should compensate inventory and order when payment fails")
        void shouldCompensateInventoryAndOrderWhenPaymentFails() {
            when(orderService.createOrder(anyString(), anyList(), any(BigDecimal.class)))
                    .thenReturn(testOrder);
            when(inventoryService.reserveInventory(anyString(), anyList()))
                    .thenReturn("reservation-789");
            when(paymentService.processPayment(anyString(), anyString(), any(BigDecimal.class)))
                    .thenThrow(new RuntimeException("Payment declined"));

            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            SagaExecutionResult result = orchestrator.execute(context);

            assertThat(result.isSuccess()).isFalse();
            assertThat(result.getFailedStep()).contains("ProcessPayment");
            assertThat(result.getCompletedSteps()).containsExactly("CreateOrder", "ReserveInventory");
            assertThat(result.getCompensationResults())
                    .isPresent()
                    .get()
                    .extracting(CompensationSummary::isTriggered)
                    .isEqualTo(true);
            assertThat(result.getCompensationResults().get().getCompletedCompensations())
                    .contains("ReserveInventory", "CreateOrder");

            verify(inventoryService).releaseInventory(eq("reservation-789"));
            verify(orderService).cancelOrder(eq("order-123"));
        }

        @Test
        @DisplayName("should compensate all steps when shipping fails")
        void shouldCompensateAllStepsWhenShippingFails() {
            when(orderService.createOrder(anyString(), anyList(), any(BigDecimal.class)))
                    .thenReturn(testOrder);
            when(inventoryService.reserveInventory(anyString(), anyList()))
                    .thenReturn("reservation-789");
            when(paymentService.processPayment(anyString(), anyString(), any(BigDecimal.class)))
                    .thenReturn("txn-101");
            when(shippingService.createShipment(anyString(), anyString()))
                    .thenThrow(new RuntimeException("Shipping unavailable"));

            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            SagaExecutionResult result = orchestrator.execute(context);

            assertThat(result.isSuccess()).isFalse();
            assertThat(result.getFailedStep()).contains("ShipOrder");
            assertThat(result.getCompletedSteps()).containsExactly(
                    "CreateOrder",
                    "ReserveInventory",
                    "ProcessPayment"
            );
            assertThat(result.getCompensationResults())
                    .isPresent()
                    .get()
                    .extracting(CompensationSummary::isTriggered)
                    .isEqualTo(true);
            assertThat(result.getCompensationResults().get().getCompletedCompensations())
                    .hasSize(3);

            verify(paymentService).refundPayment(eq("txn-101"));
            verify(inventoryService).releaseInventory(eq("reservation-789"));
            verify(orderService).cancelOrder(eq("order-123"));
        }

        @Test
        @DisplayName("should handle compensation failure gracefully")
        void shouldHandleCompensationFailureGracefully() {
            when(orderService.createOrder(anyString(), anyList(), any(BigDecimal.class)))
                    .thenReturn(testOrder);
            when(inventoryService.reserveInventory(anyString(), anyList()))
                    .thenReturn("reservation-789");
            when(paymentService.processPayment(anyString(), anyString(), any(BigDecimal.class)))
                    .thenReturn("txn-101");
            when(shippingService.createShipment(anyString(), anyString()))
                    .thenThrow(new RuntimeException("Shipping unavailable"));
            doThrow(new RuntimeException("Refund failed"))
                    .when(paymentService).refundPayment(anyString());

            SagaContext context = SagaContext.create("order-123")
                    .withMetadata("customerId", "customer-456")
                    .withMetadata("items", testItems)
                    .withMetadata("totalAmount", new BigDecimal("100.00"));

            SagaExecutionResult result = orchestrator.execute(context);

            assertThat(result.isSuccess()).isFalse();
            assertThat(result.getCompensationResults())
                    .isPresent()
                    .get()
                    .extracting(CompensationSummary::isTriggered)
                    .isEqualTo(true);
            assertThat(result.getCompensationResults().get().getFailedCompensations())
                    .contains("ProcessPayment");
            // Other compensations should still complete
            assertThat(result.getCompensationResults().get().getCompletedCompensations())
                    .contains("ReserveInventory", "CreateOrder");
        }
    }

    @Nested
    @DisplayName("Orchestrator Configuration")
    class OrchestratorConfiguration {

        @Test
        @DisplayName("should return registered steps")
        void shouldReturnRegisteredSteps() {
            List<SagaStep<?>> steps = orchestrator.getSteps();

            assertThat(steps).hasSize(4);
            assertThat(steps)
                    .extracting(SagaStep::getName)
                    .containsExactly(
                            "CreateOrder",
                            "ReserveInventory",
                            "ProcessPayment",
                            "ShipOrder"
                    );
        }
    }
}
